bjorncs/jrt transport metrics

bjorncs opened the pull request:
FYI @vekterli

bjorncs committed:
Add method to create snapshot of transport metrics

bjorncs committed:
Revert "Use JrtFactory to construct jrt supervisors in configserver"

This reverts commit 87ca5c6c4159ae5a7e44417e6624bd2d0ecb073d.

bjorncs committed:
Revert "Add jrt supervisor as constructor parameter"

This reverts commit 64c2d214d0122ba491f78c7298f0bf16ea3455bd.

bjorncs committed:
Revert "Add DefaultJrtFactory as component to container clusters"

This reverts commit 8aeee26b882c6aa1e12383502cfa37bb18082b53.

bjorncs committed:
Revert "Add tracking of jrt metrics"

This reverts commit 299d205912af1b3ebdb16a1fc037dd0770288be3.

bjorncs committed:
Make TransportMetrics a singleton

bjorncs committed:
Propagate jrt transport metrics to metrics framework

bjorncs committed:
Update copyright header

havardpe commented on jrt/src/com/yahoo/jrt/TransportMetrics.java:
```diff
@@ -76,4 +78,37 @@ public String toString() {
     }
+
+    public static class Snapshot {
+        private final long tlsCertificateVerificationFailures, peerAuthorizationFailures, serverTlsConnectionsEstablished,
```
prefer untangling members

havardpe commented on jrt/src/com/yahoo/jrt/TransportMetrics.java:
```diff
@@ -67,6 +73,15 @@ void incrementClientUnencryptedConnectionsEstablished() {
         clientUnencryptedConnectionsEstablished.incrementAndGet();
     }
 
+    void reset() {
```
maybe remove this?

havardpe commented on jrt/tests/com/yahoo/jrt/EchoTest.java:
```diff
@@ -34,33 +35,35 @@
     @Before
     public void setUp() throws ListenFailedException {
+        metrics =  TransportMetrics.getInstance();
+        metrics.reset();
```
use snapshot instead?

havardpe commented on container-disc/src/main/java/com/yahoo/container/jdisc/metric/JrtMetrics.java:
```diff
@@ -0,0 +1,44 @@
+
+    private final TransportMetrics transportMetrics = TransportMetrics.getInstance();
+    private final Metric metric;
+    private volatile Snapshot previousSnapshot;
```
either this should not need to be volatile or we need more synchronization elsewhere (the pattern is expected to be local cache for single writer)

havardpe reviewed (approved):
some minor comments

bjorncs committed:
Add calculation of metrics changes since previous snapshot

- Move metric diff calculation to TransportMetrics.Snapshot.
- Remove TransportMetrics.reset() and use changesSince() instead in EchoTest.
- Remove unnecessary volatile modifier on JrtMetrics.previousSnapshot.
- Initialize JrtMetrics.previousSnapshot in constructor.
- Use separate field declarations for fields in TransportMetrics.Snapshot.

bjorncs commented:
PTAL

havardpe reviewed (approved):
👍

bjorncs commented:
Rebased with master
